velox/expression/tests/ExprOptimizerTest.cpp

```diff
@@ -0,0 +1,396 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+
+#include "velox/common/base/tests/GTestUtils.h"
+#include "velox/expression/ExprOptimizer.h"
+#include "velox/expression/ExprRewriteRegistry.h"
+#include "velox/functions/prestosql/registration/RegistrationFunctions.h"
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+#include "velox/functions/prestosql/types/JsonType.h"
+#include "velox/functions/prestosql/types/TimestampWithTimeZoneType.h"
+
+namespace facebook::velox::expression {
+namespace {
+
+class ExprOptimizerTest : public functions::test::FunctionBaseTest {
+ protected:
+  static void SetUpTestCase() {
+    memory::MemoryManager::testingSetInstance(memory::MemoryManager::Options{});
+  }
+
+  void SetUp() override {
+    functions::prestosql::registerAllScalarFunctions("");
+    parse::registerTypeResolver();
+    options_.parseDecimalAsDouble = false;
+  }
+
+  void TearDown() override {
+    expression::ExprRewriteRegistry::instance().clear();
+  }
+
+  /// Validates expression::optimize API. The input expression is optimized and
+  /// then evaluated with a randomly generated input vector.
+  /// @param input Input SQL expression to be optimized.
+  /// @param expected Expected SQL expression that is equivalent to the result
+  /// of optimizing `input` when `failResult` is 'false', expected Velox
+  /// string representation of optimized expression otherwise.
+  /// @param type Type of `input`.
+  /// @param lambdaInputTypes Input types for any lambda sub-expressions in
+  /// `input`.
+  /// @param expectedLambdaInputTypes Input types for any lambda sub-expressions
+  /// in `expected`.
+  /// @param failResult Replace failed subexpression optimizations with a fail
+  /// expression.
+  /// @param evalError Expected error message upon evaluating `input`,
+  /// `failResult` should be true for validating error.
+  void testExpression(
+      const std::string& input,
+      const std::string& expected,
+      const RowTypePtr& type = ROW({}),
+      const std::vector<TypePtr>& lambdaInputTypes = {},
+      const std::vector<TypePtr>& expectedLambdaInputTypes = {},
+      bool failResult = false,
+      const std::string& evalError = "") {
+    const auto typedExpr = makeTypedExpr(input, type, lambdaInputTypes);
+    auto optimized =
+        expression::optimize(typedExpr, queryCtx_.get(), pool(), failResult);
+    // Generate random values of expression's type with fuzzer to validate
+    // evaluation of optimized expression.
+    const auto data = fuzzFlat(type);
+
+    if (failResult) {
+      // 'fail' expression is not supported by DuckDB's parser, hence validate
```

Comment:
> 'fail' expression is not supported by DuckDB's parser

What does this mean? "fail" is just a function, no?

Reply:
Updated, thanks.